runtime/interpreter/interpreter.go

```diff
@@ -921,6 +922,33 @@ func (interpreter *Interpreter) declareVariable(identifier string, value Value)
 	return variable
 }
 
+// declareSelfVariable declares a special "self" variable in the latest scope
+func (interpreter *Interpreter) declareSelfVariable(value Value) Variable {
```

Comment:
Use this method in places where the self variable get declared

Reply:
I think there are a couple places this PR misses where `self` is created that probably should be updated to use this function:

* https://github.com/onflow/cadence/blob/2bedb730926930bc77970bde22c13079e79d11c3/runtime/interpreter/value.go#L18276 where we create the `self` value in an attachment function.
* https://github.com/onflow/cadence/blob/2bedb730926930bc77970bde22c13079e79d11c3/runtime/interpreter/value_function.go#L355 where bound functions create `self` values

Reply:
For attachments, `self` is already a reference, so we don't really need to create another reference. We can use `declareSelfVariable` there as well, but it will internally calls the normal `declareVariable` function (because it is already a reference).

For bound functions, that place is where the 'explicit reference' for the receiver gets created for reference tracking (i.e: the existing solution for bound functions, on which this solution is also based-on).

Reply:
Ah, okay nice.